Reload apiserver certificates from disk when they change

jackkleeman opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This PR is intended to allow apiserver serving certificates and client CA to be reloaded from disk. We do this based on a file watch. This will make it it easier to rotate these certificates without any downtime. 

1. Instead of reading cert files into `*tls.Certificate` objects, instead generate a `GetConfigForClient` function which reads config from an `atomic.Value`
2. Reload the `tls.Config` synchronously on first call, and then subsequently whenever a channel fires indicating that the config is stale, because of a filewatch update on any constituent file of the config.

**Which issue(s) this PR fixes**:
Fixes #66448

**Special notes for your reviewer**:
This approach was discussed briefly in the June 12th 2019 sig auth meeting - watching that could be useful context.

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
Reload apiserver serving certificates when they change on disk.
```

jackkleeman commented:
/assign @sttts

liggitt commented:
/assign

roycaihw commented:
/sig auth

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/serving.go:
```diff
@@ -222,47 +222,85 @@ func (s *SecureServingOptions) ApplyTo(config **server.SecureServingInfo) error
+		}
+
+		// load main cert
+		serverCertFile, serverKeyFile := s.ServerCert.CertKey.CertFile, s.ServerCert.CertKey.KeyFile
```
this needs to follow a similar approach to the bearer token handling in client go:

1. at startup, synchronously load the referenced files (startup should still fail if the files cannot be read/parsed)
2. periodically (either on a time interval or potentially in response to a signal) attempt to reread the files
3. request handling uses the last successfully read values if there is an error later reading the files

liggitt replied:
the simplest would probably be for the re-read to construct a *tls.Config and on success `Store()` it in an [atomic.Value](https://golang.org/pkg/sync/atomic/#Value), then GetConfigForClient could just `return configHolder.Load().(*tls.Config)`, without needing to lock

jackkleeman replied:
@liggitt thanks for your suggestions. I have updated the branch. I kept with a time interval over a signal for now

jackkleeman commented:
/assign @liggitt

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/serving_with_loopback.go:
```diff
@@ -71,7 +71,19 @@ func (s *SecureServingOptionsWithLoopback) ApplyTo(secureServingInfo **server.Se
+			}
+
+			config.NameToCertificate[server.LoopbackClientServerNameOverride] = &tlsCert
+			config.Certificates = append(config.Certificates, tlsCert)
```
does this get called on every connection? is this modifying the cached *tls.Config that was returned for the previous connection?

jackkleeman replied:
Sadly yes. It's tricky to maintain the existing situation where this function modifies the config, given we are moving to a callback approach. Some ideas:
1) Cache here as well
2) In the function where we build the tls.Config, iterate over some list of callbacks that is appended to in this function, which mutates the config

I am interested to hear your thoughts on how we could solve this

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/serving_with_loopback.go:
```diff
@@ -71,7 +71,19 @@ func (s *SecureServingOptionsWithLoopback) ApplyTo(secureServingInfo **server.Se
 
 	default:
 		*loopbackClientConfig = secureLoopbackClientConfig
-		(*secureServingInfo).SNICerts[server.LoopbackClientServerNameOverride] = &tlsCert
```
if GetConfigForClient rebuilds from SNICerts, can we leave this as-is, adding server.LoopbackClientServerNameOverride to the SNICerts map, and let GetConfigForClient rebuild from it?

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/serving.go:
```diff
@@ -219,33 +226,82 @@ func (s *SecureServingOptions) ApplyTo(config **server.SecureServingInfo) error
+		return err
+	}
+
+	currentConfig.Store(initialConfig)
```
if we build/store the initial config in Serve(), it can honor the SNICert change made by the loopback ApplyTo call

jackkleeman commented:
Thanks @liggitt, I had been trying to think of a way to solve the wrapping callback issue.

liggitt commented:
/ok-to-test

dims commented:
/priority important-longterm
/milestone v1.16

xmudrii commented:
@jackkleeman @liggitt Hello! I'm bug triage lead for the 1.16 release cycle and considering this PR has not been updated for some time, I'd like to check what's the status of this PR. The code freeze is starting 29th August (about 2.5 weeks from now) and while there is plenty of time, we want to ensure that each PR has a chance to be merged on time.

As the PR is tagged for 1.16, is it still planned for this release?

jackkleeman commented:
@xmudrii I would love to get this merged! I defer to @liggitt

liggitt commented:
This will slip to 1.17. I'll have review bandwidth after 1.16 freeze. I think we can likely do better than a 15-minute poll.

/milestone v1.17

jackkleeman commented:
/remove-label area/codegeneration
/remove-label area/dependency

jackkleeman commented:
/assign @liggitt
I have updated to avoid a poll and instead use a filewatch, and rebased

liggitt commented:
> I have updated to avoid a poll and instead use a filewatch, and rebased

Thanks, will try to take a look this week.

cc @deads2k as relevant to https://github.com/kubernetes/kubernetes/pull/82371

liggitt commented:
I'm in favor of plumbing dynamic serving certs, and I'm ok with using that dynamic plumbing to serve updated certs when they are successfully re-read from the statically configured files. @deads2k said he would take review of this as complementary to https://github.com/kubernetes/kubernetes/issues/82141

/unassign
/assign @deads2k

jackkleeman committed:
Reload apiserver certificates and CA from disk

We use a GetConfigForClient callback to allow for dynamic generation
of *tls.Config objects for the apiserver TLS listener, and move all the
logic which generates config from files and flags into this callback,
watching all files for changes.

deads2k commented:
> I'll rebase on #83555 when merged, and try to use the dynamic cert controller

Thanks.

Can you try to structure these as a new kind of CertKeyProvider kind of like this:

```go
// CertKeyProvider provides ca bundle byte content
type CertKeyProvider interface {
	// Name is just an identifier
	Name() string
	// CurrentCertKeyProviderContent provides ca bundle byte content
	CurrentCertKeyProviderContent() ([]byte, []byte)
}
```

With a static content equivalent. I think that would be a good way to wire the plumbing and then we can add a dynamic variant.

jackkleeman commented:
Closing in favour of #83580
